mosaic-runtime/src/commonMain/kotlin/com/jakewharton/mosaic/mosaic.kt

```diff
@@ -213,28 +212,19 @@ internal class MosaicComposition(
 
 	private fun registerSnapshotApplyObserver(): ObserverHandle {
 		return Snapshot.registerApplyObserver { changedStates, _ ->
-			for (state in changedStates) {
-				if (needLayout && needDraw) {
-					break
-				}
-				if (!needLayout && readingStatesOnLayout.contains(state)) {
-					needLayout = true
-				}
-				if (!needDraw && readingStatesOnDraw.contains(state)) {
-					needDraw = true
-				}
-			}
-		}
-	}
-
-	private fun startListeningToNeedToLayoutOrDraw() {
-		scope.launch {
-			while (true) {
-				withFrameNanos {
-					when {
-						recomposer.currentState.value != Recomposer.State.Idle -> return@withFrameNanos
-						needLayout -> performLayout(applier.root)
-						needDraw -> performDraw(applier.root)
+			if (!needLayout) {
+				var setDraw = needDraw
+				for (state in changedStates) {
+					if (state in readingStatesOnLayout) {
+						needLayout = true
+						break
+					}
+					if (setDraw) {
+						continue
+					}
+					if (state in readingStatesOnDraw) {
+						setDraw = true
+						needDraw = true
```

Comment:
This should be functionally equivalent, but with early-exits where possible and minimizing volatile reads and writes.